06_puppy/nickolee/mapStore.go

```diff
@@ -0,0 +1,46 @@
+package puppystorer
+
+import "fmt"
+
+// IncrementCounter increases the ID counter everytime a new Puppy is created to prevent overwrite issues
+// in DeletePuppy()
+func (ms *MapStore) IncrementCounter() {
+	ms.counter++
+}
+
+// CreatePuppy lets you create a new unique puppy in MapStore
+func (ms *MapStore) CreatePuppy(puppy *Puppy) int {
+	ms.IncrementCounter()
+	puppy.ID = ms.counter
+	ms.store[puppy.ID] = *puppy // store the puppy object at this key in the map store (like a row in a table)
+	return puppy.ID             // return u a handle to the correct row in the table that is MapStore
+}
+
+// ReadPuppy lets you GET a puppy from MapStore if it exists. Else it will return an error
+func (ms *MapStore) ReadPuppy(id int) (*Puppy, error) {
+	// if the key exists, then puppy is assigned the value stored under the key
+	// referred to: https://blog.golang.org/go-maps-in-action
+	if puppy, ok := ms.store[id]; ok {
+		return &puppy, nil
+	}
+	// else return nil pointer to puppy and an error
+	return nil, fmt.Errorf("puppy with ID: %d does not exist", id)
+}
+
+// UpdatePuppy lets you update a "row" in MapStore
+func (ms *MapStore) UpdatePuppy(id int, puppy *Puppy) error {
+	if _, ok := ms.store[id]; !ok {
```

Comment:
what if puppy.ID and id don't match? you should add at least the error check for this, better just use Puppy. No need to pass in ID twice.

Reply:
I talked with @nicko-lee about this. We decided to ignore the id in the puppy. This allows code like
```
s.UpdatePuppy(1000, &Puppy{Name: "Fido", Breed: "Poodle", Value: 1234})
```

This feels like it should work, whereas erroring out on id clashes will cause this to throw an error.